Add blog article about cgroup v2 (#47342)
kubernetes/website: content/en/blog/_posts/2024-10-01-migrate-cgroup-v2.md

---
layout: blog
title: 'The Shift to cgroup v2 in Kubernetes: What You Need to Know'
date: 2024-10-01

> slug: kubernetes-cgroup-v2-shift

Reviewer: it should be `cgroups` plural, not a singular `cgroup`
https://en.wikipedia.org/wiki/Cgroups

applies to a number of places in the doc.

Author: Does this mean that all cgroup should be cgroups?

Reviewer 2: It should be `cgroup` according to kernel docs https://docs.kernel.org/admin-guide/cgroup-v2.html

author: >
   Paco Xu (DaoCloud)
---

`cgroups` (control groups) are a Linux kernel feature used for managing system resources.
Kubernetes uses cgroups to allocate resources like CPU and memory to containers,
ensuring that applications run smoothly without interfering with each other.

> With the release of Kubernetes v1.31, cgroup v1 has been moved into maintenance mode.

Reviewer: it's not very clear in the blog post what "maintenance mode" means.
maybe add a sentence or two after this one that explains it. in a new paragraph.

Author: The details are in another blog.
https://github.com/pacoxu/website/blob/b8bddc4aff4e9f67a35957dd07acce7b3d0ebdea/content/en/blog/_posts/2024-08-14-moving-cgroup-v1-support-maintenance-mode-kubernetes-1-31.md#L69-L76

Could we just add a link here?

For cgroup v2, it graduated in v1.25 2 years ago.

Top FAQs are why we should migrate, what's the benifits and lost,
and what needs to be noticed when using cgroup v2.

> ## cgroup v1 problem, and solutions in cgroup v2

Reviewer: ```suggestion
## cgroups v1 problems, and solutions in cgroups v2
```

Cgroup v1 and cgroup official doc can be found in

- [v1 doc](https://www.kernel.org/doc/Documentation/cgroup-v1/)
- [v2 doc](https://www.kernel.org/doc/Documentation/cgroup-v2.txt)

> Let's talk from the known issue.

Reviewer: ```suggestion
Let's list some known issues.
```

is that better?

### active_file memory is not considered as available memory

There is [a known issue](/docs/concepts/scheduling-eviction/node-pressure-eviction/#active-file-memory-is-not-considered-as-available-memory) of page cache: [#43916](https://github.com/kubernetes/kubernetes/issues/43916).

- In cgroup v1, we have no native solutions.
  Workarounds are setting larger memory limit for Pods or using some external projects to drop cache or
  throttling memory allocating when memory is beyond a threshold.
- In cgroup v2, we can use `memory.high` to throttle.

Support for Memory QoS was initially added in Kubernetes v1.22, 
and later some limitations around the formula for calculating `memory.high` were identified.
These limitations are addressed in Kubernetes v1.27.

However, until v1.31, the feature gate is still alpha due to another known issue
that application pod may be hanging forever due to heavy memory reclaiming.

> ### container aware OOM killer and better OOM handling strategies

Reviewer: ```suggestion
### Container aware OOM killer and better OOM handling strategies
```

In cgroup v2, one process of a multi-processes Pod could be killed by the OOM killer.
In this case, Pod has to use [runit](https://github.com/void-linux/runit) or
supervisord to manage multi processes lifecycle.

Cgroup v2 uses `cgroup.kill` file.
Writing “1” to the file causes the cgroup and all descendant cgroups to be killed.
This means that all processes located in the affected cgroup tree will be killed via SIGKILL.
Pod may run multiple processes, and all processes can be killed simultaneously.

As mentioned above, cgroup v2 `memory.high` can throttle the new memory allocation and
cgroup can be aware of the OOM earsiler.

> Besides, PSI can also help to know memory load. [oomd](https://github.com/facebookincubator/oomd) is a good example

Reviewer: ```suggestion
Besides, PSI can also help to know the memory load. [oomd](https://github.com/facebookincubator/oomd) is a good example
```

using PSI to implement a userspace out-of-memory killer.

> ### rootless support

Reviewer: ```suggestion
### root-less support
```
vs
```suggestion
### Rootless support
```

Author: I use `Rootless`.

- In 4.5, cgroup v2 io,memory & pids were supported.
- Kernel 4.15 supports CPU controller for cgroup v2
- PSI support starts from kernel 4.20.
- Using cgroup v2 in kernel older than 5.2 is not recommended due to lack of freezer.
- In Kubernetes, 5.8 is the minimal kernel version for cgroup v2 as root `cpu.stat` file on cgroupv2
  was only added on kernel 5.8.
- `memory.peak` is added in 5.19.

### Use systemd as cgroup driver

[Configure the kubelet's cgroup driver to match the container runtime cgroup driver](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/).

The [Container runtimes](/docs/setup/production-environment/container-runtimes)
page explains that the `systemd` driver is recommended for kubeadm
based setups instead of the kubelet's default `cgroupfs` driver,
because kubeadm manages the kubelet as a systemd service.

A minimal example of configuring the field explicitly:

```YAML
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
cgroupDriver: systemd
```

In v1.31, [KEP-4033](https://github.com/kubernetes/enhancements/issues/4033) is beta to extend CRI API for the kubelet

> to discover the cgroup driver from the container runtime. This will help installer and kubelet to 

Author: Yes. I still want to add more information here and some more paragraphs.

> ### Troubleshooting Tools and Commands

Reviewer 3: I recommend:
```suggestion
### Tools and commands for troubleshooting
```
